File: utilz/family.py
from __future__ import annotations

from .func import Route, setDictOptional
from typing import TYPE_CHECKING
from . import utils
import discord
### -> Exceptions <- ###
from .errors import (
  utilsError,
  ResponseError
)

if TYPE_CHECKING:
  from .guild import Guild

def fixData(data: dict) -> dict:
  if "role" in data:
    data["role"] = str(data["role"].id if isinstance(data["role"], discord.Role) else data["role"])
  if "roles" in data:
    for i, item in enumerate(data["roles"]):
      data["roles"][i] = str(item.id if isinstance(item, discord.Role) else item)
  if "habilitys" in data:
    data["habilitys"] = [str(i.id) for i in data["habilitys"]]
  return data

class FamilyAlreadyExists(utilsError):
  def __init__(self, family: Family) -> None:
    super().__init__(f"Family id  {family.id} already exists!")
class Family:
  def __init__(
    self,
    Guild: Guild,
    data: dict[str, dict]
  ) -> None:
    self.ep = Guild
    
    self.name: str = data["name"]
    self.id: int = int(data["id"])
    self.roleID: utils.Optional[int] = int(data["role"])
    self.rolesID: utils.Optional[list[int]] = [int(i) for i in data["roles"]]
    self.limit: utils.Optional[int] = data["limit"]
    self.require: int = data["require"]
    self.rarity: int = data["rarity"]
    self.embed_title: str = data["embed"]["title"]
    self.embed_description: str = data["embed"]["description"]
    self.embed_image: utils.Optional[str] = data["embed"]["url"]
  def __repr__(self) -> str:
    return f"utilz.Family(name={self.name!r} id={self.id} role={self.role})"
  def __eq__(self, __other: utils.Union[discord.Role, str, int]) -> bool:
    if isinstance(__other, Family):
      return self.id == __other.id
    __other = (__other.id if isinstance(__other, discord.Role) else __other)
    if not isinstance(__other, str) and not isinstance(__other, int):
      return False
    return (
      __other == self.roleID
      or __other == self.id
      or __other == self.name
    )
  @property
  def role(self) -> utils.Optional[discord.Role]:
    return self.ep._ep.get_role(self.roleID)
  @property
  def roles(self) -> utils.Optional[list[discord.Role]]:
    return (None if self.rolesID is None else [self.ep._ep.get_role(i) for i in self.roleID])
  def get_data(self) -> dict[str, dict]:
    response = Route.get(f"/desktop/Guilds/{self.ep.id}/Family/{self.id}")
    if not response.status_code == 200:
      raise ResponseError(f"Failhed connection family id {self.id} of guild id {self.ep.id}")
    return response.json()
  def verificationUser(self, User) -> bool:
    verificationLimit = (self.limit is None)
    verificationRoles = bool(len(self.rolesID))
    verificationInventory = User.rolls_familys.toJson().get(str(self.id), 0) < 3
    if not verificationInventory:
      return False
    if not verificationLimit:
      players = self.ep.users.fromFamily(self)
      verificationLimit = (players is None or not len(players) >= self.limit)
      if not verificationLimit:
        return False
    if not verificationRoles:
      roles = [i for i in User.auth.roles if i.id in self.rolesID]
      verificationRoles = len(roles) >= self.require
      if not verificationRoles:
        return False
    return (verificationLimit and verificationRoles and verificationInventory)
  def embed(self, **kwgs) -> discord.Embed:
    title: str = kwgs.get("title", self.embed_title)
    description: str = kwgs.get("description", self.embed_description)
    url: str = kwgs.get("url", self.embed_image)
    embed = discord.Embed(
      title=title,
      description=description,
      color=0x71368A
    )
    if not url is None:
      embed.set_image(url=url)
    return embed
  def edit(self, **kwgs) -> Family: 
    data = setDictOptional({}, "name", kwgs.get("name"))
    data = setDictOptional(data, "role", kwgs.get("role"))
    data = setDictOptional(data, "roles", kwgs.get("roles"))
    data = setDictOptional(data, "habilitys", kwgs.get("habilitys"))
    data = setDictOptional(data, "rarity", kwgs.get("rarity"))
    data = setDictOptional(data, "require", kwgs.get("require"))
    data = setDictOptional(data, "limit", kwgs.get("limit"))
    embed = setDictOptional({}, "title", kwgs.get("embed_title"))
    embed = setDictOptional(embed, "description", kwgs.get("embed_description"))
    embed = setDictOptional(embed, "url", kwgs.get("embed_url"))
    message = setDictOptional({}, "content", kwgs.get("message_content"))
    message = setDictOptional(message, "url", kwgs.get("message_url"))
    if len(embed):
      data["embed"] = embed
    if len(message):
      data["message"] = message
    if not len(data):
      return self
    response = Route.patch(
      f"/desktop/Guilds/{self.ep.id}/Family/{self.id}",
      data=fixData(data),
      headers={
        "content-type": "application/json"
      }
    )
    if not response.status_code == 200:
      raise ResponseError(f"Failed connection in family id {self.id} of guild id {self.ep.id}")
    self.__init__(self.ep, response.json())
    return self
def New_Family(
  Guild: Guild,
  name: str,
  role: discord.Role,
  **kwgs
) -> Family:
  response = Route.post(
    f"/desktop/Guilds/{Guild.id}/Family",
    data=fixData({
      "name": name,
      "role": str(role.id)
    }),
    headers={
      "content-type": "application/json"
    }
  ); switch = utils.Switch(response.status_code)
  if switch(404):
    raise ResponseError(f"Failhed connection of guild id {Guild.id}")
  return Family(
    Guild=Guild,
    data=response.json()
  ).edit(**kwgs)

File: server/utils/Family.ts
import { Json, JsonArray, embed } from "./types"
import { writeFileSync } from 'fs'
import { Guild } from "./Guild"
import { getitem } from "./utils"
import path from 'path'

let drr = path.normalize(path.join(__dirname, '../', '../', "/Json"))

interface FamilyDDO {
  name: string,
  id: string|number,
  role: string|number|null,
  roles: string[],
  habilitys: string[],
  rarity: number,
  require: number,
  limit: number|null,
  embed: embed,
  message: {
    content: string|null,
    url: string|null
  }
}
let FamilyData: FamilyDDO = {
  name: "NAME",
  id: "5500550",
  role: null,
  roles: [],
  habilitys: [],
  rarity: 0,
  require: 1,
  limit: null,
  embed: {
    title: null,
    description: null,
    url: null
  },
  message: {
    content: null,
    url: null
  }
}

function fixData(data: Json, family: FamilyDDO): FamilyDDO {
  data["name"] = getitem(data, "name", family["name"])
  data["id"] = family["id"]
  data["role"] = getitem(data, "role", family["role"])
  data["roles"] = getitem(data, "roles", family["roles"])
  data["habilitys"] = getitem(data, "habilitys", family["habilitys"])
  data["rarity"] = getitem(data, "rarity", family["rarity"])
  data["require"] = getitem(data, "require", family["require"])
  data["limit"] = getitem(data, "limit", family["limit"])
  if(data["embed"] !== undefined) {
    data["embed"] = {
      title: getitem(data["embed"], "title", family["embed"]["title"]),
      description: getitem(data["embed"], "description", family["embed"]["description"]),
      url: getitem(data["embed"], "url", family["embed"]["url"])
    }
  } else {data["embed"] = family["embed"]}
  if(data["message"] !== undefined) {
    data["message"] = {
      content: getitem(data["message"], "content", family["message"]["content"]),
      url: getitem(data["message"], "url", family["message"]["url"])
    } 
  } else {data["message"] = family["message"]}; return {
    name: data["name"],
    id: data["id"],
    role: data["role"],
    roles: data["roles"],
    habilitys: data["habilitys"],
    rarity: data["rarity"],
    require: data["require"],
    limit: data["limit"],
    embed: data["embed"],
    message: data["message"]
  }
}

export class Familys extends Array<Family> {
  public guild: Guild
  constructor(guild: Guild, all: FamilyDDO[]) {
    let callable = () => {
      let array: Family[] = []
      for(let elem of all) {
        array.push(new Family(elem))
      }; return array
    }; super(...callable())
    this.guild = guild
  }
  public toJson(): FamilyDDO[] {
    let callable = (): FamilyDDO[] => {
      let array: FamilyDDO[] = []
      for(let family of this) {
        array.push(family.family_data)
      }; return array
    }; return callable()
  }
  public get(id: string): Family|null {
    for(let family of this) {
      if(family.id == id) {return family;}
    }; return null
  }
  public new(name: string, opts: Json): Family {
    opts["name"] = name
    FamilyData["id"] = Math.floor(Math.random() * 100000000000000000).toString()
    let data = fixData(opts, FamilyData)
    this.push(new Family(data))
    this.save()
    return this[this.length-1]
  }
  public save(): void {
    let fileName = `${drr}/Family/${this.guild.id}.json`
    let array: FamilyDDO[] = this.toJson()
    writeFileSync(fileName, JSON.stringify(array, null, 2), {encoding: 'utf-8'})
  }
}
export class Family {
  public id: string|number;
  public family_data: FamilyDDO;
  constructor(data: FamilyDDO) {
    this.family_data = data
    this.id = data["id"]
  }
  public edit(family: Json): Family {
    this.family_data = fixData(family, this.family_data)
    return this
  }
}